src/Compiler/Utilities/illib.fs

```diff
@@ -118,25 +118,51 @@ module internal PervasiveAutoOpens =
 
             task.Result
 
-/// An efficient lazy for inline storage in a class type. Results in fewer thunks.
-[<Struct>]
-type InlineDelayInit<'T when 'T: not struct> =
-    new(f: unit -> 'T) =
-        {
-            store = Unchecked.defaultof<'T>
-            func = Func<_>(f)
-        }
+[<AbstractClass>]
+type DelayInitArrayMap<'T, 'TDictKey, 'TDictValue>(f: unit -> 'T[]) =
+    let syncObj = obj ()
+
+    let mutable arrayStore = null
+    let mutable dictStore = null
+
+    let mutable func = f
+
+    member this.GetArray() =
+        match arrayStore with
+        | NonNull value -> value
+        | _ ->
+            Monitor.Enter(syncObj)
+
+            try
+                match arrayStore with
+                | NonNull value -> value
+                | _ ->
+
+                    arrayStore <- func ()
 
-    val mutable store: 'T
-    val mutable func: Func<'T> MaybeNull
+                    func <- Unchecked.defaultof<_>
+                    arrayStore
+            finally
+                Monitor.Exit(syncObj)
 
-    member x.Value =
-        match x.func with
-        | null -> x.store
+    member this.GetDictionary() =
+        match dictStore with
+        | NonNull value -> value
         | _ ->
-            let res = LazyInitializer.EnsureInitialized(&x.store, x.func)
-            x.func <- null
-            res
+            let array = this.GetArray()
+            Monitor.Enter(syncObj)
```

Comment:
Shouldn't the two resulting values (array vs dict) each have its own synchronization object?
They do not have to wait for each other, or?

Reply:
> They do not have to wait for each other, or?

They do. The dictionary is created from the array.

Reply:
OK